Source/WebKit/WebProcess/WebPage/WebContextMenu.cpp

```diff
@@ -63,10 +63,15 @@ void WebContextMenu::show()
     RefPtr<API::Object> userData;
     menuItemsWithUserData(menuItems, userData);
 
-    auto menuLocation = view->contentsToRootView(controller.hitTestResult().roundedPointInInnerNodeFrame());
+    auto& hitTestResult = controller.hitTestResult();
+    auto menuLocation = view->contentsToRootView(hitTestResult.roundedPointInInnerNodeFrame());
 
     ContextMenuContextData contextMenuContextData(menuLocation, menuItems, controller.context());
 
+    auto imageURL = hitTestResult.absoluteImageURL();
+    if (!imageURL.isEmpty() && !hitTestResult.imageRect().isEmpty() && m_page->hasLocalDataForURL(imageURL))
+        contextMenuContextData.setHasEntireImage(true);
```

Comment:
I think it might make more sense to move this logic into the `ContextMenuContextData` constructor, which has access to the `HitTestResult`.

Reply:
I considered that, but `ContextMenuContextData` doesn't have access to the page.

Reply:
You could get it from the `HitTestResult`, `innerNodeFrame()->page()`.

Reply:
That's actually the `WebCore::Page`, and  `hasLocalDataForURL()` is a `WebKit::WebPage` function. But I still think you're right that the logic makes more sense in `ContextMenuContextData`, so I'm going to add a `WebPage&` to the constructor.

Reply:
I think that would be incorrect since `ContextMenuContextData` is in `WebKit/Shared`, but `WebPage` is a `WebProcess` only concept.

How about we move the implementation of `hasLocalDataForURL()` to `WebCore::Page` instead?

Reply:
That sounds reasonable. Assuming I can do that, do you think it also makes sense to move the calculation of `hasEntireImage` into `WebCore::ContextMenuContext` as well and just have `WebKit::ContextMenuData` initialize its member variable from `ContextMenuContext`?

Reply:
Sounds good to me!

Reply:
Actually, now that I think about it, we could just put everything on `HitTestResult` / `WebHitTestResultData` / `_WKHitTestResult`, since this just operates on hit test result data.

The reason the QR code payload is a separate property it that it cannot be determined at hit test result time.

Reply:
@whsieh @hortont424 Any thoughts on whether this SPI should live on `_WKHitTestResult` or `_WKContextMenuElementInfo`? Today, it will only be used for context menus, but in reality it's simply derived from properties on `_WKHitTestResult` and could be generally useful.

Reply:
I left the property on `_WKContextMenuElementInfo` for now, but I'll leave this conversation open to get other people's opinions.